modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java

```diff
@@ -0,0 +1,338 @@
+     */
+    @Test
+    public void testLostLearners() throws Exception {
+        RaftNode follower = nodes.get(0);
+        List<RaftNode> learners = nodes.subList(1, nodes.size());
+
+        List<CompletableFuture<RaftGroupService>> services = nodes.stream()
+                .map(node -> startRaftGroup(node, new TestRaftGroupListener(), List.of(follower), learners))
+                .collect(toList());
+
+        // Wait for the leader to be elected.
+        // Wait for the leader to be elected.
+        services.forEach(service -> {
+            assertThat(service.thenCompose(s -> s.refreshLeader().thenApply(v -> s.leader())), willBe(follower.asPeer()));
+        });
+
+        nodes.set(1, null).close();
+        nodes.set(2, null).close();
+
+        assertThat(services.get(0).thenCompose(RaftGroupService::refreshLeader), willCompleteSuccessfully());
+    }
+
+    private CompletableFuture<RaftGroupService> startRaftGroup(
+            RaftNode raftNode,
+            RaftGroupListener listener,
+            List<RaftNode> peers,
+            List<RaftNode> learners
+    ) {
+        try {
+            CompletableFuture<RaftGroupService> future = raftNode.loza.prepareRaftGroup(
+                    RAFT_GROUP_NAME,
+                    peers.stream().map(RaftNode::localMember).collect(toList()),
+                    learners.stream().map(RaftNode::localMember).collect(toList()),
+                    () -> listener,
+                    () -> noopLsnr,
+                    RaftGroupOptions.defaults()
+            );
+
+            return future.thenApply(s -> {
+                s.timeout(100);
+
+                return s;
+            });
+        } catch (NodeStoppingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static class TestRaftGroupListener implements RaftGroupListener {
+        final BlockingQueue<String> storage = new LinkedBlockingQueue<>();
+
+        @Override
+        public void onWrite(Iterator<CommandClosure<WriteCommand>> iterator) {
+            iterator.forEachRemaining(closure -> {
+                assertThat(closure.command(), is(instanceOf(TestWriteCommand.class)));
+
+                TestWriteCommand writeCommand = (TestWriteCommand) closure.command();
+
+                if (!storage.contains(writeCommand.value)) {
+                    storage.add(writeCommand.value);
+                }
+
+                closure.result(null);
+            });
+        }
+
+        @Override
+        public void onRead(Iterator<CommandClosure<ReadCommand>> iterator) {
+        }
+
+        @Override
+        public void onSnapshotSave(Path path, Consumer<Throwable> doneClo) {
+        }
+
+        @Override
+        public boolean onSnapshotLoad(Path path) {
+            return true;
+        }
+
+        @Override
+        public void onShutdown() {
+        }
+    }
+
+    private static class TestWriteCommand implements WriteCommand {
```

Comment:
You know what, can you make it a network message in advance? We're porting all commands right now, so you'll save us some time